librdmacm/rsocket.c

```diff
@@ -1454,7 +1454,7 @@ static int rs_do_connect(struct rsocket *rs)
 		rs->state = rs_connect_rdwr;
 		break;
 	default:
-		ret = ERR(EINVAL);
+		ret = (rs->state & rs_connected) ? 0 : ERR(EINVAL);
```

Comment:
Isn't this code racy?
We can set rs_connected state inside rs_accept() and that function is not protected by lock.

Reply:
This code is done under lock, so wouldn't say this change is racy.  It may be that other code is.

rs_accept() is a little different, in that the application does not have access to the rsocket being modified there until raccept returns.  rs_accept() is only called from the cm thread, and only that thread has access to the rsocket until after rsocket is placed on the accept queue.

Reply:
To clarify, there are races in code like this (in rrecv):
```
	if (rs->state & rs_opening) {
		ret = rs_do_connect(rs);
```
The state is checked outside of a lock.  By the time rs_do_connect() runs, the rs->state has changed (to connected in this case) by the CM thread.  rs_do_connect repeats the state check, but under lock; however, it incorrectly assumes that the rsocket cannot already be connected.

Reply:
Thanks for the clarification.

Indeed, I meant that race is between non-locked access to rs_connected and this line.